### PR TITLE
graphql: Generic data loaders

### PR DESCRIPTION
**Pull request by efritz**
Replace preloader/bulk loader in sentine/codenav with a unified interface.

## Test plan

Existing resolvers still work.

### CODE REVIEW

File: enterprise/internal/codeintel/shared/resolvers/dataloader.go [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+	"context"
+	"sync"
+)
+
+type Identifiable[T any] interface {
+	GetID() T
+}
+
+type MultiFactory[K, V any] interface {
+	Create(ctx context.Context, id K) ([]V, error)
```
Comment (Numbers88s) [inline enterprise/internal/codeintel/shared/resolvers/dataloader.go:13]: I would name it something closer to `fetch` or `fetchByID`

File: enterprise/internal/codeintel/shared/types/upload.go [hunk 2]
```diff
@@ -29,6 +29,9 @@ type Upload struct {
 	ShouldReindex     bool
 }
 
+func (u Upload) GetID() int {
+	return u.ID
+}
```
Comment (eseliger) [inline enterprise/internal/codeintel/shared/types/upload.go:34]: should we standardize on RecordID? Seems easy enough to rename later if needed,but reduces the amount of methods one needs to implement

File: enterprise/internal/codeintel/shared/resolvers/dataloader.go [hunk 3]
```diff
@@ -0,0 +1,117 @@
+package sharedresolvers
+
+import (
+	"context"
+	"sync"
+)
+
+type Identifiable[T any] interface {
```
Comment (eseliger) [inline enterprise/internal/codeintel/shared/resolvers/dataloader.go:8]:
```suggestion
type GetIDer[T any] interface {
```
😆

File: enterprise/internal/codeintel/shared/resolvers/dataloader.go (hunk 1 above)
Comment (eseliger) [reply enterprise/internal/codeintel/shared/resolvers/dataloader.go:13]: ➕

File: enterprise/internal/codeintel/shared/resolvers/dataloader.go [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+	dl := &DataLoader[K, V]{
+		svc: svc,
+		ids: map[K]struct{}{},
+	}
+
+	dl.cache = NewDoubleLockedCache[K, V](MultiFactoryFunc[K, V](dl.load))
+	return dl
+}
+
+func (l *DataLoader[K, V]) Mark(ids ...K) {
```
Comment (eseliger) [inline enterprise/internal/codeintel/shared/resolvers/dataloader.go:88]: Mark who?

Find the name a bit non-self-explanatory, so maybe try something like or whatever

```suggestion
func (l *DataLoader[K, V]) Presubmit(ids ...K) {
```

**Review by eseliger — APPROVED** [review]
LGTM, would be cool to have this outside of the codeintel codebase.

File: enterprise/internal/codeintel/shared/resolvers/dataloader.go (hunk 3 above)
Comment (efritz) [reply enterprise/internal/codeintel/shared/resolvers/dataloader.go]: Kinda done :)

**Pull request merged**